### PR TITLE
Deploy and proposal for USDT market on Goerli

### PR DESCRIPTION
**Pull request by cwang25**
This PR introduces the deployment for USDT Goerli as well as the proposal to initialize the market there.

TODO:

- [ ]  deploy USDT instance to Goerli

- [ ] run market initialization proposal

### CODE REVIEW

**Comment by cwang25** [comment]
The scenario will fail as we encountered back in Arbitrum native usdc deploy. 
Since this is deploying another comet on the chain with existing governor contract, without proper setFactory, setConfigurator on migration script, the scenario tests will run into errors. 

But this can be confirmed in migration PR, which with migration script the scenario run will succeed.

File: contracts/Comet.sol
Comment (cwang25) [inline contracts/Comet.sol:1]:
```
 ···································|··············|·················
 |  CometFactory                    ·      24.044  ·                │
 ···································|··············|·················
 |  CometHarness                    ·      20.606  ·                │
 ···································|··············|·················
 |  CometMath                       ·       0.062  ·                │
 ···································|··············|·················
 |  CometModified                   ·      18.977  ·                │
 ···································|··············|·················
 |  CometModifiedFactory            ·      24.182  ·                │
 ···································|··············|·················
```

The contract size is approaching to limit, compiler will complain with `Warning: 2 contracts exceed the size limit for mainnet deployment.`
But the actual size limit is 24.576 kb, so we should be fine :P

File: contracts/Comet.sol [hunk 1]
```diff
@@ -760,18 +760,49 @@ contract Comet is CometMainInterface {
...
+                case 32 {                      // This is a compliant ERC-20
+                    returndatacopy(0, 0, 32)
+                    success := mload(0)        // Set `success = returndata` of override external call
+                }
+                default {                      // This is an excessively non-compliant ERC-20, revert.
+                    revert(0, 0)
+                }
+        }
         if (!success) revert TransferInFailed();
+        return IERC20NonStandard(asset).balanceOf(address(this)) - preTransferBalance;
```
Comment (kevincheng96) [inline contracts/Comet.sol:784]: This return value doesn't seem to be used anywhere right now, meaning fee tokens are not properly handled. You'll want to use the returned amount for accounting.

File: contracts/Comet.sol [hunk 2]
```diff
@@ -760,18 +760,49 @@ contract Comet is CometMainInterface {
...
     }
 
     /**
      * @dev Safe ERC20 transfer out
      */
```
Comment (kevincheng96) [inline contracts/Comet.sol:789]: Can you also add the dev note here:

`     * @dev Note: Safely handles non-standard ERC-20 tokens that do not return a value. See here: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
`

File: scenario/SupplyScenario.ts [hunk 3]
```diff
@@ -136,6 +137,42 @@ scenario(
...
+  {
+    tokenBalances: {
+      albert: { $base: 1000 }, // in units of asset, not wei
+    },
+    filter: async (ctx) => matchesDeployment(ctx, [{ deployment: 'usdt' }])
+  },
+  async ({ comet, actors }, context, world) => {
+    // Set fees for USDT for testing
+    const USDTAdmin = await world.deploymentManager.getSigner();
+    const USDT = await world.deploymentManager.existing('USDT', await comet.baseToken(), 'goerli');
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:151]: You can use `world.base.network` to programmatically fetch the network instead of hardcoding it.

File: scenario/SupplyScenario.ts [hunk 4]
```diff
@@ -136,6 +137,42 @@ scenario(
...
+    await USDT.connect(USDTAdmin).setParams(10, 10);
+
+    const { albert } = actors;
+    const baseAssetAddress = await comet.baseToken();
+    const baseAsset = context.getAssetByAddress(baseAssetAddress);
+    const scale = (await comet.baseScale()).toBigInt();
+
+    expect(await baseAsset.balanceOf(albert.address)).to.be.equal(1000n * scale);
+
+    // Albert supplies 100 units of base to Comet
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:162]:
```suggestion
    // Albert supplies 1000 units of base to Comet
```

File: scenario/SupplyScenario.ts [hunk 5]
```diff
@@ -167,6 +204,45 @@ scenario(
...
+    },
+    cometBalances: {
+      albert: { $base: -1000 } // in units of asset, not wei
+    },
+    filter: async (ctx) => matchesDeployment(ctx, [{ deployment: 'usdt' }]),
+  },
+  async ({ comet, actors }, context, world) => {
+    // Set fees for USDT for testing
+    const USDTAdmin = await world.deploymentManager.getSigner();
+    const USDT = await world.deploymentManager.existing('USDT', await comet.baseToken(), 'goerli');
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:221]: Same here

File: scenario/SupplyScenario.ts [hunk 6]
```diff
@@ -167,6 +204,45 @@ scenario(
...
+    const { albert } = actors;
+    const baseAssetAddress = await comet.baseToken();
+    const baseAsset = context.getAssetByAddress(baseAssetAddress);
+    const scale = (await comet.baseScale()).toBigInt();
+    const utilization = await comet.getUtilization();
+    const borrowRate = (await comet.getBorrowRate(utilization)).toBigInt();
+
+    expectApproximately(await albert.getCometBaseBalance(), -1000n * scale, getInterest(1000n * scale, borrowRate, 1n) + 1n);
+
+    // Albert repays 100 units of base borrow
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:234]:
```suggestion
    // Albert repays 1000 units of base borrow
```

File: scenario/SupplyScenario.ts [hunk 7]
```diff
@@ -167,6 +204,45 @@ scenario(
...
+    await baseAsset.approve(albert, comet.address);
+    const txn = await albert.supplyAsset({ asset: baseAsset.address, amount: 1000n * scale });
+
+    // XXX all these timings are crazy
+    // Expect to have -1000000, due to token fee, alber only repay 999 USDT instead of 1000 USDT, thus alber still owe 1 USDT which is 1000000
+    expectApproximately(await albert.getCometBaseBalance(), -1000000n, getInterest(1000n * scale, borrowRate, 4n) + 2n);
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:240]: Can you use `exp(1, 6)` instead? The number of 0s is hard to read

Comment (kevincheng96) [inline scenario/SupplyScenario.ts:240]: Do you need `expectApproximately`? Does exact expect work?

File: scenario/SupplyScenario.ts [hunk 8]
```diff
@@ -200,6 +276,46 @@ scenario(
...
+    const baseAsset = context.getAssetByAddress(baseAssetAddress);
+    const scale = (await comet.baseScale()).toBigInt();
+
+    expect(await baseAsset.balanceOf(albert.address)).to.be.equal(1000n * scale);
+    expect(await comet.balanceOf(betty.address)).to.be.equal(0n);
+
+    await baseAsset.approve(albert, comet.address);
+    await albert.allow(betty, true);
+
+    // Betty supplies 100 units of base from Albert
```
Comment (kevincheng96) [inline scenario/SupplyScenario.ts:305]:
```suggestion
    // Betty supplies 1000 units of base from Albert
```

**Review by kevincheng96 — COMMENTED** [review]
Looks great! Could you also add unit tests for `supply base`, `supply collateral`, and `buy collateral`? I notice there is no scenarios for `buy collateral`, so would be good to get some unit test coverage at least.

File: scenario/SupplyScenario.ts (hunk 3 above)
Comment (cwang25) [reply scenario/SupplyScenario.ts:151]: Oh i see. Thank you so much! 😄

Comment (cwang25) [reply scenario/SupplyScenario.ts]: Replaced all with world.base.network

File: scenario/SupplyScenario.ts (hunk 7 above)
Comment (cwang25) [reply scenario/SupplyScenario.ts]: Changed to exp(1,6) but since it's negative so I added  `-1n *`

Comment (cwang25) [reply scenario/SupplyScenario.ts]: I think in test cases without fee token also has to use expectApproximately() 🤔 
I tried use exact will get 2wei diff.

File: contracts/Comet.sol (hunk 2 above)
Comment (cwang25) [reply contracts/Comet.sol:790]: added to doTransferOut

File: deployments/goerli/usdt/configuration.json [hunk 9]
```diff
@@ -0,0 +1,54 @@
+{
+  "name": "Compound USDT",
+  "symbol": "cUSDTv3",
+  "baseToken": "USDT",
+  "baseTokenPriceFeed": "0xAb5c49580294Aff77670F839ea425f5b78ab3Ae7",
+  "borrowMin": "10e6",
```
Comment (kevincheng96) [inline deployments/goerli/usdt/configuration.json:6]: Let's make this 1e0

File: deployments/goerli/usdt/configuration.json [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    "borrowSlopeLow": 0.035,
+    "borrowSlopeHigh": 0.25,
+    "borrowBase": 0.015
+  },
+  "rewardToken": "COMP",
+  "tracking": {
+    "indexScale": "1e15",
+    "baseSupplySpeed": "0.000402083333333e15",
+    "baseBorrowSpeed": "0.000402083333333e15",
+    "baseMinForRewards": "10000e6"
```
Comment (kevincheng96) [inline deployments/goerli/usdt/configuration.json:26]: Let's lower this to 500.

File: deployments/goerli/usdt/configuration.json [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    "borrowKink": 0.8,
+    "borrowSlopeLow": 0.035,
+    "borrowSlopeHigh": 0.25,
+    "borrowBase": 0.015
+  },
+  "rewardToken": "COMP",
+  "tracking": {
+    "indexScale": "1e15",
+    "baseSupplySpeed": "0.000402083333333e15",
+    "baseBorrowSpeed": "0.000402083333333e15",
```
Comment (kevincheng96) [inline deployments/goerli/usdt/configuration.json:25]: Should be 0 right?

File: deployments/goerli/usdt/deploy.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  // Send some forked USDT to timelock
+  await deploymentManager.idempotent(
+    async () => await USDT.connect(signer).balanceOf(timelock.address) == 0,
+    async () => {
+      trace(`Sending USDC to timelock`);
+      await USDT.connect(signer).transfer(
+        timelock.address,
+        exp(50_000_000, 6),
+      );
+      trace(`Sent USDC to timelock completed`);
```
Comment (kevincheng96) [inline deployments/goerli/usdt/deploy.ts:43]: Should say USDT instead of USDC

File: deployments/goerli/usdt/deploy.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+  // Import shared contracts from cUSDCv3
+  const cometAdmin = await deploymentManager.fromDep('cometAdmin', 'goerli', 'usdc');
+  // Purposely don't use the factory because Comet implementation changed.
+  // const cometFactory = await deploymentManager.fromDep('cometFactory', 'goerli', 'usdc');
+  const $configuratorImpl = await deploymentManager.fromDep('configurator:implementation', 'goerli', 'usdc');
+  const configurator = await deploymentManager.fromDep('configurator', 'goerli', 'usdc');
+  const rewards = await deploymentManager.fromDep('rewards', 'goerli', 'usdc');
+  const fauceteer = await deploymentManager.fromDep('fauceteer', 'goerli', 'usdc');
+  const fxRoot = await deploymentManager.fromDep('fxRoot', 'goerli', 'usdc');
```
Comment (kevincheng96) [inline deployments/goerli/usdt/deploy.ts:29]: I don't think you need this, it's only for Polygon

File: deployments/goerli/usdt/relations.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+export default {
+  ...baseRelationConfig,
+  fxRoot: {
+    relations: {
+      stateSender: {
+        field: async fxRoot => fxRoot.stateSender()
+      }
+    }
+  },
```
Comment (kevincheng96) [inline deployments/goerli/usdt/relations.ts:11]: Same here

File: deployments/goerli/usdt/deploy.ts (hunk 12 above)
Comment (cwang25) [reply deployments/goerli/usdt/deploy.ts]: Fixed.

File: deployments/goerli/usdt/deploy.ts (hunk 13 above)
Comment (cwang25) [reply deployments/goerli/usdt/deploy.ts]: removed fxRoot.

File: deployments/goerli/usdt/relations.ts (hunk 14 above)
Comment (cwang25) [reply deployments/goerli/usdt/relations.ts]: removed :)

File: deployments/goerli/usdt/configuration.json (hunk 11 above)
Comment (cwang25) [reply deployments/goerli/usdt/configuration.json]: set to 0. :)

File: deployments/goerli/usdt/configuration.json (hunk 10 above)
Comment (cwang25) [reply deployments/goerli/usdt/configuration.json]: 500e6 👌

File: deployments/goerli/usdt/configuration.json (hunk 9 above)
Comment (cwang25) [reply deployments/goerli/usdt/configuration.json]: set to 1e0 1wei.

**Comment by cwang25** [comment]
> Looks great! Could you also add unit tests for `supply base`, `supply collateral`, and `buy collateral`? I notice there is no scenarios for `buy collateral`, so would be good to get some unit test coverage at least.

I have pushed the unit test for all 3 cases :)

File: contracts/test/NonStandardFaucetFeeToken.sol [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+ *  See https://github.com/ethereum/EIPs/issues/20
+ * @dev With USDT fee token mechanism
+ * @dev Note: `transfer` and `transferFrom` do not return a boolean
+ */
+contract NonStandardFeeToken {
+    string public name;
+    string public symbol;
+    uint8 public decimals;
+    uint256 public totalSupply;
+    mapping (address => mapping (address => uint256)) public allowance;
```
Comment (kevincheng96) [inline contracts/test/NonStandardFaucetFeeToken.sol:16]:
```suggestion
    mapping(address => mapping (address => uint256)) public allowance;
```

File: contracts/test/NonStandardFaucetFeeToken.sol [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+pragma solidity 0.8.15;
+
+/**
+ * @title Non-standard ERC20 token
+ * @dev Implementation of the basic standard token.
+ *  See https://github.com/ethereum/EIPs/issues/20
+ * @dev With USDT fee token mechanism
+ * @dev Note: `transfer` and `transferFrom` do not return a boolean
+ */
+contract NonStandardFeeToken {
```
Comment (kevincheng96) [inline contracts/test/NonStandardFaucetFeeToken.sol:11]: Can you make this extend the `IERC20NonStandard` just to make sure the interfaces match?

File: contracts/test/NonStandardFaucetFeeToken.sol [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+            balanceOf[address(this)] = balanceOf[address(this)] + fee;
+        }
+
+        allowance[src][msg.sender] = allowance[src][msg.sender] - amount;
+        balanceOf[src] = balanceOf[src] - amount;
+        balanceOf[dst] = balanceOf[dst] + sendAmount;
+        emit Transfer(src, dst, sendAmount);
+    }
+
+    function approve(address _spender, uint256 amount) external returns (bool) {
```
Comment (kevincheng96) [inline contracts/test/NonStandardFaucetFeeToken.sol:72]: This returns a `bool` even though the non-standard version doesn't.

File: test/supply-test.ts [hunk 18]
```diff
@@ -405,8 +405,127 @@ describe('supplyTo', function () {
...
+
+  it('supplies collateral the correct amount in a fee-like situation', async () => {
+    const assets = defaultAssets();
+    // Add FeeToken Collateral to assets on top of default assets
+    assets['FeeToken'] = {
+      initial: 1e8,
+      decimals: 18,
+      factory: (await ethers.getContractFactory('NonStandardFaucetFeeToken')) as NonStandardFaucetFeeToken__factory,
+    };
+
```
Comment (kevincheng96) [inline test/supply-test.ts:480]: nit: extra line

File: contracts/Comet.sol
Comment (kevincheng96) [inline contracts/Comet.sol:1]: For withdraw, should the `Withdraw` event emit the original amount or the amount after fees?

Comment (kevincheng96) [reply contracts/Comet.sol:1]: I think probably just the original amount.

File: test/buy-collateral-test.ts [hunk 19]
```diff
@@ -323,8 +323,89 @@ describe('buyCollateral', function () {
...
+        from: alice.address,
+        to: comet.address,
+        amount: exp(49.5, 6),
+      }
+    });
+    expect(event(txn, 1)).to.be.deep.equal({
+      Transfer: {
+        from: comet.address,
+        to: alice.address,
+        amount: 54450000000000000000n,
```
Comment (kevincheng96) [inline test/buy-collateral-test.ts:398]: You can use `exp` here.

File: test/buy-collateral-test.ts [hunk 20]
```diff
@@ -323,8 +323,89 @@ describe('buyCollateral', function () {
...
+      }
+    });
+    expect(event(txn, 2)).to.be.deep.equal({
+      BuyCollateral: {
+        buyer: alice.address,
+        asset: COMP.address,
+        baseAmount: exp(49.5, 6),
+        collateralAmount: 55000000000000000000n,
```
Comment (kevincheng96) [inline test/buy-collateral-test.ts:406]: Same here.

File: test/buy-collateral-test.ts [hunk 21]
```diff
@@ -323,8 +323,89 @@ describe('buyCollateral', function () {
...
+    const txn = await wait(cometAsA.buyCollateral(COMP.address, exp(50, 18), 50e6, alice.address));
+    const p1 = await portfolio(protocol, alice.address);
+    const r1 = await comet.getReserves();
+
+    expect(r0).to.be.equal(0n);
+    expect(r0).to.be.lt(await comet.targetReserves());
+    expect(p0.internal).to.be.deep.equal({ USDT: 0n, COMP: 0n });
+    expect(p0.external).to.be.deep.equal({ USDT: exp(100, 6), COMP: 0n });
+    expect(p1.internal).to.be.deep.equal({ USDT: 0n, COMP: 0n });
+    expect(p1.external).to.be.deep.equal({ USDT: exp(50, 6), COMP: 54450000000000000000n });
```
Comment (kevincheng96) [inline test/buy-collateral-test.ts:385]: `exp` here as well

**Review by kevincheng96 — COMMENTED** [review]
Nice work on the tests! They look great 👍

File: contracts/test/NonStandardFaucetFeeToken.sol (hunk 16 above)
Comment (cwang25) [reply contracts/test/NonStandardFaucetFeeToken.sol]: np 👍

File: contracts/test/NonStandardFaucetFeeToken.sol (hunk 17 above)
Comment (cwang25) [reply contracts/test/NonStandardFaucetFeeToken.sol]: oh yeah,  I have now removed the return bool statement :)

File: test/supply-test.ts (hunk 18 above)
Comment (cwang25) [reply test/supply-test.ts:480]: removed 👍

File: contracts/Comet.sol
Comment (cwang25) [reply contracts/Comet.sol:1]: Yeah this probably boils down to UX.
If 1% fee, user deposit 1000 token to contract, user will see balance show as 999.  (In a sense contract receives only 999 token).  If user try to withdraw 999 from contract, user only receive 998.1 🤔  (In a sense contract gives up 999 token)
If based on contract's perspective, then I think original amount make more sense to me.

Also, I just tried to add another balance compare in transferout to return calculated amount. 
(contract price might go up to... 24.218 and  24.355 still lower than the absolute max limit)

File: contracts/test/NonStandardFaucetFeeToken.sol [hunk 22]
```diff
@@ -1,14 +1,15 @@
 // SPDX-License-Identifier: BUSL-1.1
 pragma solidity 0.8.15;
 
+import "../IERC20NonStandard.sol";
```
Comment (kevincheng96) [inline contracts/test/NonStandardFaucetFeeToken.sol:4]: nit: can you add an extra line below

**Review by kevincheng96 — APPROVED** [review]
Think this should be ready for audit with OZ now!

Could you update the PR description to say this introduces a new implementation of Comet that supports non-standard ERC20s and fee tokens? 

Separately, I wonder if we should add this logic in a new `Comet` file, like `NonStandardComet` or `FeeTokenComet`?

**Comment by kevincheng96** [comment]
Looking at the re-entrancy test fixes, I think the better fix is actually just to have the `transfer` and `transferFrom` functions in the `EvilToken` do actual transfer accounting while also calling `performAttack()`.

**Comment by cwang25** [comment]
Synced with ⛑️  offline. 
For the re-entrancy issue that Comet is not immune to the tokens that's got exploited. (i.e. EvilToken.sol) 
We decided to not adding additional layer to prevent that in Comet as this vulnerability will require governance to be exploited that governance has to vote and add the suspicious token into Comet's contract. If suspicious token contract is not added to Comet's market, this should not be a concern. 


For reference, if import re-entrancy guard from openzeppelin which can prevent exploited token contract to execute re-entrancy attack, but it will also increase the contract size to: 
```
 ···································|··············|·················
 |  CometFactory                    ·      24.173  ·                │
 ···································|··············|·················
 |  CometModifiedFactory            ·      24.311  ·                │
 ···································|··············|·················
 ```

File: contracts/ReentrancyGuard.sol [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+
+    // The values being non-zero value makes deployment a bit more expensive,
+    // but in exchange the refund on every call to nonReentrant will be lower in
+    // amount. Since refunds are capped to a percentage of the total
+    // transaction's gas, it is best to keep them low in cases like this one, to
+    // increase the likelihood of the full refund coming into effect.
+    uint256 private constant _NOT_ENTERED = 1;
+    uint256 private constant _ENTERED = 2;
+
+    uint256 private _status;
```
Comment (kevincheng96) [inline contracts/ReentrancyGuard.sol:37]: Does this clash with existing storage variables if we exist an existing Comet to it? We need to be very careful about introducing new storage slots

Comment (cwang25) [reply contracts/ReentrancyGuard.sol]: Ah yea it does. :( 
I just removed this import. 
I added those field following the CometStorage.sol. 
```
Add const STATE append to the end of the consts in CometCore's constant field
Add state variable append to the end of the variables in CometStorage
Add customer error append to the end of the custom errors in CometMainInterface
Add modifier functions in Comet.sol
```

Hope this will resolve the clash issue.

Comment (cwang25) [reply contracts/ReentrancyGuard.sol]: Discussed offline, we will not add the reentrancyguard to comet.

File: test/supply-test.ts [hunk 24]
```diff
@@ -436,10 +554,11 @@ describe('supplyTo', function () {
     await EVIL.setAttack(attack);
 
     await comet.connect(alice).allow(EVIL.address, true);
-
+    await wait(EVIL.connect(alice).approve(comet.address, 75e6));
+    await EVIL.allocateTo(alice.address, 75e6);
     await expect(
       comet.connect(alice).supplyTo(bob.address, EVIL.address, 75e6)
-    ).to.be.revertedWith("custom error 'SupplyCapExceeded()'");
+    ).to.be.revertedWith("custom error 'TransferInFailed()'");
```
Comment (cwang25) [inline test/supply-test.ts:561]: Now the accounting in comet won't be inflated by re-entrancy supply reaching SupplyCapExceeded(). 
It will be blocked and throw TransferInFailed() error.

File: test/withdraw-test.ts [hunk 25]
```diff
@@ -511,7 +511,7 @@ describe('withdraw', function () {
       // in callback, EVIL token calls transferFrom(alice.address, bob.address, 1e6)
       await expect(
         comet.connect(alice).withdraw(EVIL.address, 1e6)
-      ).to.be.revertedWith("custom error 'NotCollateralized()'");
+      ).to.be.revertedWith("custom error 'TransferOutFailed()'");
```
Comment (cwang25) [inline test/withdraw-test.ts:514]: Now the accounting in comet won't be mistaken by re-entrancy withdraw reaching NotColalteralized() error.
Instead, it will be blocked and throw TransferInFailed() error right away.

File: contracts/Comet.sol [hunk 26]
```diff
@@ -201,6 +201,32 @@ contract Comet is CometMainInterface {
         (asset14_a, asset14_b) = getPackedAssetInternal(config.assetConfigs, 14);
     }
 
+    modifier nonReentrant() {
+        _nonReentrantBefore();
+        _;
+        _nonReentrantAfter();
```
Comment (kevincheng96) [inline contracts/Comet.sol:207]: Let's inline these

File: contracts/Comet.sol [hunk 27]
```diff
@@ -201,6 +201,32 @@ contract Comet is CometMainInterface {
...
+    }
+
+    function _nonReentrantBefore() private {
+        bytes32 slot = COMET_REENTRANCY_GUARD_FLAG_SLOT;
+        uint256 status;
+        assembly {
+            status := sload(slot)
+        }
+
+        if(status == COMET_REENTRANCY_GUARD_ENTERED) revert ReentrantCallBlocked();
```
Comment (kevincheng96) [inline contracts/Comet.sol:217]: nit: space after `if`

File: contracts/CometCore.sol [hunk 28]
```diff
@@ -56,6 +56,13 @@ abstract contract CometCore is CometConfiguration, CometStorage, CometMath {
     /// @dev The scale for factors
     uint64 internal constant FACTOR_SCALE = 1e18;
 
+    /// @dev The storage slot for reentrancy guard flags
+    bytes32 internal constant COMET_REENTRANCY_GUARD_FLAG_SLOT = bytes32(keccak256("comet.reentrancy.guard"));
+
+    /// @dev The reentrancy guard status
+    uint256 internal constant COMET_REENTRANCY_GUARD_NOT_ENTERED = 1;
+    uint256 internal constant COMET_REENTRANCY_GUARD_ENTERED = 2;
+
```
Comment (kevincheng96) [inline contracts/CometCore.sol:65]: Can just inline rather than creating constants. Also, let's use a default value of 0. Otherwise, you need to initialize the slot to 1.

File: contracts/Comet.sol [hunk 29]
```diff
@@ -776,7 +802,7 @@ contract Comet is CometMainInterface {
      * @dev Safe ERC20 transfer out
      * @dev Note: Safely handles non-standard ERC-20 tokens that do not return a value. See here: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
      */
-    function doTransferOut(address asset, address to, uint amount) internal {
+    function doTransferOut(address asset, address to, uint amount) nonReentrant internal {
```
Comment (kevincheng96) [inline contracts/Comet.sol:805]: Having `nonReentrant` on the transfer functions breaks `BuyCollateral` right?

File: test/buy-collateral-test.ts [hunk 30]
```diff
@@ -543,8 +543,8 @@ describe('buyCollateral', function () {
       expect(normalTotalsBasic.trackingSupplyIndex).to.equal(evilTotalsBasic.trackingSupplyIndex);
       expect(normalTotalsBasic.trackingBorrowIndex).to.equal(evilTotalsBasic.trackingBorrowIndex);
       expect(normalTotalsBasic.totalSupplyBase).to.equal(1e6);
-      // EvilToken attack is using 3000 EVIL tokens, so totalSupplyBase should have 3000e6
-      expect(evilTotalsBasic.totalSupplyBase).to.equal(3000e6);
+      // EvilToken attack should be blocked
+      expect(evilTotalsBasic.totalSupplyBase).to.equal(0);
```
Comment (kevincheng96) [inline test/buy-collateral-test.ts:547]: Let's add an assertion to assert the re-entrancy error.

File: test/supply-test.ts [hunk 31]
```diff
@@ -558,47 +558,7 @@ describe('supplyTo', function () {
...
-    ).to.be.revertedWith("custom error 'SupplyCapExceeded()'");
-  });
-
-  // This test is purposely test the edge case where comet internal accounting will be incorrect when EvilToken re-entrancy attack
-  // is performed. The attack will cause comet to have inflated supply amount of token. 
-  //
-  // We decided that this won't be a concern for Comet, as in order to pull off this attack, the governance has to propose and vote to 
-  // add suspicious token in to Comet's market. As long as governance doesn't add suspicious token contract or erc-777 token to market, the 
-  // Comet should not be vulnerable to this type of attack.
-  it('incorrect accounting via re-entrancy', async () => {
```
Comment (kevincheng96) [inline test/supply-test.ts]: Instead of removing this test, let's re-purpose it to show that re-entrancy is no longer possible.

File: test/withdraw-test.ts [hunk 32]
```diff
@@ -558,7 +558,7 @@ describe('withdraw', function () {
       // in callback, EvilToken attempts to withdraw USDC to bob's address
       await expect(
         comet.connect(alice).withdraw(EVIL.address, 1e6)
-      ).to.be.revertedWith("custom error 'NotCollateralized()'");
+      ).to.be.revertedWith("custom error 'TransferOutFailed()'");
```
Comment (kevincheng96) [inline test/withdraw-test.ts:561]: Test name should change as well

File: test/withdraw-test.ts (hunk 25 above)
Comment (kevincheng96) [reply test/withdraw-test.ts:514]: Shouldn't it throw the re-entrancy error instead?

File: contracts/Comet.sol (hunk 26 above)
Comment (cwang25) [reply contracts/Comet.sol:207]: ok will remove the modifier function, and inline the functions into TransferIn and TransferOut.

File: contracts/CometCore.sol (hunk 28 above)
Comment (cwang25) [reply contracts/CometCore.sol:65]: I recalled seeing in openzeppelin's reentrancy guard contract saying that using 1, 2 is cheaper on running than setting it to 0.  (reference: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/ReentrancyGuard.sol) 

Since we're only checking on if status == COMET_REENTRANCY_GUARD_ENTERED.  It looks like the initialization value of 0 or 1 wouldn't affect the first run behavior, so I ended up changing to 1,2.  And each run it only checks on if the status is ENTERED(2) :P.   But yeah I agree the first initialization value having to be 0 instead of 1 is a bit confusing, and I can change it back to 0 if the gas saving gain is not enough to tradeoff with the confusion initialization values.

File: test/supply-test.ts (hunk 31 above)
Comment (cwang25) [reply test/supply-test.ts]: Sure no problem. 
The reason I removed it, it's because I found other old tests (that assumed re-entrancy can be done by exceeding supply cap) sort of already re-purposed and covered the block of re-entrancy already.

File: test/withdraw-test.ts (hunk 25 above)
Comment (cwang25) [reply test/withdraw-test.ts:514]: Yeah it was bit weird,  but since the error is thrown in the nested transfer() function, so it gets sort of "re-wrap" into TransferInFailed() as the code here doesn't re-raise the error (or should we make it re-raise the error?) 
```
uint256 preTransferBalance = ERC20(asset).balanceOf(address(this));
        (bool success, bytes memory returndata) = asset.call(abi.encodeWithSelector(ERC20.transferFrom.selector, from, address(this), amount));
        if (!success || !(returndata.length == 0 || abi.decode(returndata, (bool)))) {
            revert TransferInFailed();
        }
        return ERC20(asset).balanceOf(address(this)) - preTransferBalance;
```

From code we can see now the transferFrom revert won't reveirt right away with returndata, but instead when it fails, we throw `TransferInFailed()`.
